[Bug] [Worker] Match wrong pid on linux server

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.


### What happened

If you deploy DS on linux server, cancel process instance, then it's possible to get error log:
```
[ERROR] 2023-07-27 23:14:54.887 +0800 org.apache.dolphinscheduler.server.worker.processor.WorkerTaskKillProcessor:[196] - kill task error
org.apache.dolphinscheduler.common.shell.AbstractShell$ExitCodeException: kill: (26483): Operation not permitted
kill: (558): No such process
kill: (1493): No such process
```

### What you expected to happen

No error log.

### How to reproduce

Ditto.

### Anything else

That is because in `ProcessUtils.getPidsStr`, only match pids on macos and windows.

```java
        if (SystemUtils.IS_OS_MAC) {
            String pids = OSUtils.exeCmd(String.format("%s -sp %d", TaskConstants.PSTREE, processId));
            if (null != pids) {
                mat = MACPATTERN.matcher(pids);
            }
        } else {
            String pids = OSUtils.exeCmd(String.format("%s -p %d", TaskConstants.PSTREE, processId));
            mat = WINDOWSATTERN.matcher(pids);
        }
```

If run on linux server, will get pids string like: `sudo(6279)---558_1497.sh(6282)---sleep(6354)`, but if use windows matcher,  it will match "6279 558 1497 6282 6354", expected matching results is "6279 6282 6354".

### Version

dev

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtils.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtils.java
@@ -76,7 +76,12 @@ public final class ProcessUtils {
     /**
      * Expression of PID recognition in Windows scene
      */
-    private static final Pattern WINDOWSATTERN = Pattern.compile("(\\d+)");
+    private static final Pattern WINDOWSPATTERN = Pattern.compile("(\\d+)");
+
+    /**
+     * Expression of PID recognition in Linux scene
+     */
+    private static final Pattern LINUXPATTERN = Pattern.compile("\\((\\d+)\\)");
 
     /**
      * kill tasks according to different task types.
@@ -117,12 +122,19 @@ public final class ProcessUtils {
         // pstree pid get sub pids
         if (SystemUtils.IS_OS_MAC) {
             String pids = OSUtils.exeCmd(String.format("%s -sp %d", TaskConstants.PSTREE, processId));
-            if (null != pids) {
+            if (StringUtils.isNotEmpty(pids)) {
                 mat = MACPATTERN.matcher(pids);
+            }
+        } else if (SystemUtils.IS_OS_LINUX) {
+            String pids = OSUtils.exeCmd(String.format("%s -p %d", TaskConstants.PSTREE, processId));
+            if (StringUtils.isNotEmpty(pids)) {
+                mat = LINUXPATTERN.matcher(pids);
             }
         } else {
             String pids = OSUtils.exeCmd(String.format("%s -p %d", TaskConstants.PSTREE, processId));
-            mat = WINDOWSATTERN.matcher(pids);
+            if (StringUtils.isNotEmpty(pids)) {
+                mat = WINDOWSPATTERN.matcher(pids);
+            }
         }
 
         if (null != mat) {
@@ -218,5 +230,4 @@ public final class ProcessUtils {
                 .getPodLogWatcher(
                         new KubernetesApplicationManagerContext(k8sTaskExecutionContext, taskAppId, containerName));
     }
-
 }

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtilsTest.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtilsTest.java
@@ -17,14 +17,13 @@
 
 package org.apache.dolphinscheduler.plugin.task.api.utils;
 
-
 import org.apache.commons.lang3.StringUtils;
-import org.junit.jupiter.api.Assertions;
-import org.junit.jupiter.api.Test;
 
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
 
 public class ProcessUtilsTest {
 
@@ -33,7 +32,7 @@ public class ProcessUtilsTest {
     private String getPidStr(String pids) {
         Matcher mat = null;
         StringBuilder sb = new StringBuilder();
-        if(StringUtils.isNotEmpty(pids)){
+        if (StringUtils.isNotEmpty(pids)) {
             mat = LINUXPATTERN.matcher(pids);
         }
         if (null != mat) {

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtilsTest.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtilsTest.java
@@ -77,7 +77,7 @@ public class ProcessUtilsTest {
             command3 = String.format("%s -p %d", TaskConstants.PSTREE, processId3);
         }
         osUtilsMockedStatic.when(() -> OSUtils.exeCmd(command3)).thenReturn(pids3);
-        String actualPidsStr3 = ProcessUtils.getPidsStr(processId2);
+        String actualPidsStr3 = ProcessUtils.getPidsStr(processId3);
         Assertions.assertEquals(exceptPidsStr3, actualPidsStr3);
     }
 

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtilsTest.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtilsTest.java
@@ -53,7 +53,6 @@ public class ProcessUtilsTest {
         int processId2 = 2000;
         String exceptPidsStr2 = "2000 2100 2101";
         String command2;
-        MockedStatic<OSUtils> osUtilsMockedStatic2 = Mockito.mockStatic(OSUtils.class);
         if (SystemUtils.IS_OS_MAC) {
             command2 = String.format("%s -sp %d", TaskConstants.PSTREE, processId2);
         } else if (SystemUtils.IS_OS_LINUX) {
@@ -61,7 +60,7 @@ public class ProcessUtilsTest {
         } else {
             command2 = String.format("%s -p %d", TaskConstants.PSTREE, processId2);
         }
-        osUtilsMockedStatic2.when(() -> OSUtils.exeCmd(command2)).thenReturn(pids2);
+        osUtilsMockedStatic.when(() -> OSUtils.exeCmd(command2)).thenReturn(pids2);
         String actualPidsStr2 = ProcessUtils.getPidsStr(processId2);
         Assertions.assertEquals(exceptPidsStr2, actualPidsStr2);
 
@@ -70,7 +69,6 @@ public class ProcessUtilsTest {
         int processId3 = 5000;
         String exceptPidsStr3 = "5000 6000 7000 7100";
         String command3;
-        MockedStatic<OSUtils> osUtilsMockedStatic3 = Mockito.mockStatic(OSUtils.class);
         if (SystemUtils.IS_OS_MAC) {
             command3 = String.format("%s -sp %d", TaskConstants.PSTREE, processId3);
         } else if (SystemUtils.IS_OS_LINUX) {
@@ -78,7 +76,7 @@ public class ProcessUtilsTest {
         } else {
             command3 = String.format("%s -p %d", TaskConstants.PSTREE, processId3);
         }
-        osUtilsMockedStatic3.when(() -> OSUtils.exeCmd(command3)).thenReturn(pids3);
+        osUtilsMockedStatic.when(() -> OSUtils.exeCmd(command3)).thenReturn(pids3);
         String actualPidsStr3 = ProcessUtils.getPidsStr(processId2);
         Assertions.assertEquals(exceptPidsStr3, actualPidsStr3);
     }

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtilsTest.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtilsTest.java
@@ -1,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.dolphinscheduler.plugin.task.api.utils;
+
+
+import org.apache.commons.lang3.StringUtils;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+
+public class ProcessUtilsTest {
+
+    private static final Pattern LINUXPATTERN = Pattern.compile("\\((\\d+)\\)");
+
+    private String getPidStr(String pids) {
+        Matcher mat = null;
+        StringBuilder sb = new StringBuilder();
+        if(StringUtils.isNotEmpty(pids)){
+            mat = LINUXPATTERN.matcher(pids);
+        }
+        if (null != mat) {
+            while (mat.find()) {
+                sb.append(mat.group(1)).append(" ");
+            }
+        }
+        return sb.toString().trim();
+    }
+
+    @Test
+    public void testGetPidsStr() {
+        String pids = "sudo(6279)---558_1497.sh(6282)---sleep(6354)";
+        String exceptPidsStr = "6279 6282 6354";
+        String actualPidsStr = getPidStr(pids);
+        Assertions.assertEquals(exceptPidsStr, actualPidsStr);
+
+        String pids2 = "init(1)---systemd(1000)---(sd-pam)(1001)";
+        String exceptPidsStr2 = "1 1000 1001";
+        String actualPidsStr2 = getPidStr(pids2);
+        Assertions.assertEquals(exceptPidsStr2, actualPidsStr2);
+
+        String pids3 = "sshd(5000)---sshd(6000)---bash(7000)---python(7100)";
+        String exceptPidsStr3 = "5000 6000 7000 7100";
+        String actualPidsStr3 = getPidStr(pids3);
+        Assertions.assertEquals(exceptPidsStr3, actualPidsStr3);
+
+        String pids4 = "apache2(2000)---apache2-submit_task.py(2100)---apache2(2101)";
+        String exceptPidsStr4 = "2000 2100 2101";
+        String actualPidsStr4 = getPidStr(pids4);
+        Assertions.assertEquals(exceptPidsStr4, actualPidsStr4);
+    }
+
+}

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtilsTest.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtilsTest.java
@@ -17,53 +17,70 @@
 
 package org.apache.dolphinscheduler.plugin.task.api.utils;
 
-import org.apache.commons.lang3.StringUtils;
+import org.apache.dolphinscheduler.common.utils.OSUtils;
+import org.apache.dolphinscheduler.plugin.task.api.TaskConstants;
 
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
+import org.apache.commons.lang3.SystemUtils;
 
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
+import org.mockito.MockedStatic;
+import org.mockito.Mockito;
 
 public class ProcessUtilsTest {
 
-    private static final Pattern LINUXPATTERN = Pattern.compile("\\((\\d+)\\)");
-
-    private String getPidStr(String pids) {
-        Matcher mat = null;
-        StringBuilder sb = new StringBuilder();
-        if (StringUtils.isNotEmpty(pids)) {
-            mat = LINUXPATTERN.matcher(pids);
-        }
-        if (null != mat) {
-            while (mat.find()) {
-                sb.append(mat.group(1)).append(" ");
-            }
-        }
-        return sb.toString().trim();
-    }
-
     @Test
-    public void testGetPidsStr() {
+    public void testGetPidsStr() throws Exception {
+        // first
         String pids = "sudo(6279)---558_1497.sh(6282)---sleep(6354)";
+        int processId = 6279;
         String exceptPidsStr = "6279 6282 6354";
-        String actualPidsStr = getPidStr(pids);
+        String command;
+        MockedStatic<OSUtils> osUtilsMockedStatic = Mockito.mockStatic(OSUtils.class);
+        if (SystemUtils.IS_OS_MAC) {
+            command = String.format("%s -sp %d", TaskConstants.PSTREE, processId);
+        } else if (SystemUtils.IS_OS_LINUX) {
+            command = String.format("%s -p %d", TaskConstants.PSTREE, processId);
+        } else {
+            command = String.format("%s -p %d", TaskConstants.PSTREE, processId);
+        }
+        osUtilsMockedStatic.when(() -> OSUtils.exeCmd(command)).thenReturn(pids);
+        String actualPidsStr = ProcessUtils.getPidsStr(processId);
         Assertions.assertEquals(exceptPidsStr, actualPidsStr);
 
-        String pids2 = "init(1)---systemd(1000)---(sd-pam)(1001)";
-        String exceptPidsStr2 = "1 1000 1001";
-        String actualPidsStr2 = getPidStr(pids2);
+        // second
+        String pids2 = "apache2(2000)---222332-apache2-submit_task.py(2100)---apache2(2101)";
+        int processId2 = 2000;
+        String exceptPidsStr2 = "2000 2100 2101";
+        String command2;
+        MockedStatic<OSUtils> osUtilsMockedStatic2 = Mockito.mockStatic(OSUtils.class);
+        if (SystemUtils.IS_OS_MAC) {
+            command2 = String.format("%s -sp %d", TaskConstants.PSTREE, processId2);
+        } else if (SystemUtils.IS_OS_LINUX) {
+            command2 = String.format("%s -p %d", TaskConstants.PSTREE, processId2);
+        } else {
+            command2 = String.format("%s -p %d", TaskConstants.PSTREE, processId2);
+        }
+        osUtilsMockedStatic2.when(() -> OSUtils.exeCmd(command2)).thenReturn(pids2);
+        String actualPidsStr2 = ProcessUtils.getPidsStr(processId2);
         Assertions.assertEquals(exceptPidsStr2, actualPidsStr2);
 
+        // Third
         String pids3 = "sshd(5000)---sshd(6000)---bash(7000)---python(7100)";
+        int processId3 = 5000;
         String exceptPidsStr3 = "5000 6000 7000 7100";
-        String actualPidsStr3 = getPidStr(pids3);
+        String command3;
+        MockedStatic<OSUtils> osUtilsMockedStatic3 = Mockito.mockStatic(OSUtils.class);
+        if (SystemUtils.IS_OS_MAC) {
+            command3 = String.format("%s -sp %d", TaskConstants.PSTREE, processId3);
+        } else if (SystemUtils.IS_OS_LINUX) {
+            command3 = String.format("%s -p %d", TaskConstants.PSTREE, processId3);
+        } else {
+            command3 = String.format("%s -p %d", TaskConstants.PSTREE, processId3);
+        }
+        osUtilsMockedStatic3.when(() -> OSUtils.exeCmd(command3)).thenReturn(pids3);
+        String actualPidsStr3 = ProcessUtils.getPidsStr(processId2);
         Assertions.assertEquals(exceptPidsStr3, actualPidsStr3);
-
-        String pids4 = "apache2(2000)---apache2-submit_task.py(2100)---apache2(2101)";
-        String exceptPidsStr4 = "2000 2100 2101";
-        String actualPidsStr4 = getPidStr(pids4);
-        Assertions.assertEquals(exceptPidsStr4, actualPidsStr4);
     }
 
 }

